File: juce_cmp_ui/lib/src/main/kotlin/juce_cmp/renderer/IOSurfaceRenderer.kt
// SPDX-FileCopyrightText: 2026 Luciano Iam <oss@lucianoiam.com>
// SPDX-License-Identifier: MIT

package juce_cmp.renderer

import androidx.compose.runtime.Composable
import androidx.compose.ui.InternalComposeUiApi
import androidx.compose.ui.graphics.asComposeCanvas
import androidx.compose.ui.scene.CanvasLayersComposeScene
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.IntSize
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.ptr.IntByReference
import kotlinx.coroutines.*
import juce_cmp.ipc.Ipc
import juce_cmp.ipc.JuceValueTree
import juce_cmp.input.InputDispatcher
import juce_cmp.input.InputEvent
import juce_cmp.input.InputType
import org.jetbrains.skia.*
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/**
 * Renders Compose content to an IOSurface using GPU-accelerated zero-copy rendering.
 *
 * @param socketFD The socket file descriptor for IPC
 * @param scaleFactor The display scale factor (e.g., 2.0 for Retina)
 * @param ipc The IPC channel for communication with host
 * @param onFrameRendered Optional callback invoked after each frame is rendered
 * @param onJuceEvent Optional callback when host sends events of type JUCE (JuceValueTree payload)
 * @param content The Compose content to render
 */
fun runIOSurfaceRenderer(
    socketFD: Int,
    scaleFactor: Float = 1f,
    ipc: Ipc,
    onFrameRendered: ((frameNumber: Long, surface: Surface) -> Unit)? = null,
    onJuceEvent: ((tree: JuceValueTree) -> Unit)? = null,
    content: @Composable () -> Unit
) {
    runIOSurfaceRendererImpl(socketFD, scaleFactor, ipc, onFrameRendered, onJuceEvent, content)
}

/**
 * Native library for zero-copy IOSurface rendering.
 *
 * Provides Metal device/queue pointers and IOSurface-backed textures.
 */
private interface NativeLib : Library {
    fun createMetalContext(): Pointer?
    fun destroyMetalContext(context: Pointer)
    fun getMetalDevice(context: Pointer): Pointer?
    fun getMetalQueue(context: Pointer): Pointer?
    fun createIOSurfaceTexture(context: Pointer, surfaceID: Int, outWidth: IntByReference?, outHeight: IntByReference?): Pointer?
    fun releaseIOSurfaceTexture(texturePtr: Pointer)
    fun flushAndSync(context: Pointer)

    companion object {
        val INSTANCE: NativeLib by lazy {
            // Extract native library from JAR resources at runtime
            val libFile = Native.extractFromResourcePath("iosurface_renderer")
            Native.load(libFile.absolutePath, NativeLib::class.java)
        }
    }
}

/**
 * Holds the Skia/Metal resources for rendering to an IOSurface.
 * These need to be recreated when the window resizes.
 */
private class RenderResources(
    val texturePtr: Pointer,
    val directContext: DirectContext,
    val skiaSurface: Surface,
    val width: Int,
    val height: Int
) : AutoCloseable {
    override fun close() {
        skiaSurface.close()
        directContext.close()
        NativeLib.INSTANCE.releaseIOSurfaceTexture(texturePtr)
    }
}

/**
 * Creates RenderResources from an IOSurface ID.
 */
private fun createRenderResources(
    metalContext: Pointer,
    devicePtr: Pointer,
    queuePtr: Pointer,
    surfaceID: Int
): RenderResources {
    val widthRef = IntByReference()
    val heightRef = IntByReference()
    val texturePtr = NativeLib.INSTANCE.createIOSurfaceTexture(
        metalContext, surfaceID, widthRef, heightRef
    ) ?: error("Failed to create IOSurface-backed texture for ID $surfaceID")

    val width = widthRef.value
    val height = heightRef.value

    // Create Skia DirectContext using our Metal device/queue
    val directContext = DirectContext.makeMetal(
        Pointer.nativeValue(devicePtr),
        Pointer.nativeValue(queuePtr)
    )

    // Create BackendRenderTarget wrapping the IOSurface-backed texture
    val renderTarget = BackendRenderTarget.makeMetal(
        width, height,
        Pointer.nativeValue(texturePtr)
    )

    // Create Skia Surface from the render target
    val skiaSurface = Surface.makeFromBackendRenderTarget(
        directContext,
        renderTarget,
        SurfaceOrigin.TOP_LEFT,
        SurfaceColorFormat.BGRA_8888,
        ColorSpace.sRGB
    ) ?: error("Failed to create Skia Surface from BackendRenderTarget")

    return RenderResources(texturePtr, directContext, skiaSurface, width, height)
}

/**
 * Zero-copy GPU-accelerated IOSurface renderer implementation.
 *
 * This implementation uses Skia's Metal backend to render Compose content
 * directly into an IOSurface-backed Metal texture. There are NO CPU pixel
 * copies - the GPU renders directly to the shared memory that the host
 * process displays.
 *
 * Architecture:
 * 1. Host creates IOSurface with kIOSurfaceIsGlobal and sends ID via socket
 * 2. Native library looks up IOSurface by ID
 * 3. Native library creates an MTLTexture backed by the IOSurface
 * 4. Skia's DirectContext.makeMetal() uses our Metal device/queue
 * 5. Skia's BackendRenderTarget.makeMetal() wraps the IOSurface texture
 * 6. Compose's CanvasLayersComposeScene renders to this surface
 * 7. GPU work goes directly to the IOSurface - host sees it immediately!
 */
@OptIn(InternalComposeUiApi::class)
private fun runIOSurfaceRendererImpl(
    socketFD: Int,
    scaleFactor: Float = 1f,
    ipc: Ipc,
    onFrameRendered: ((frameNumber: Long, surface: Surface) -> Unit)? = null,
    onJuceEvent: ((tree: JuceValueTree) -> Unit)? = null,
    content: @Composable () -> Unit
) {
    // Create Metal context (device + command queue)
    val metalContext = NativeLib.INSTANCE.createMetalContext()
        ?: error("Failed to create Metal context")

    try {
        // Get Metal device and queue pointers for Skia
        val devicePtr = NativeLib.INSTANCE.getMetalDevice(metalContext)
            ?: error("Failed to get Metal device")
        val queuePtr = NativeLib.INSTANCE.getMetalQueue(metalContext)
            ?: error("Failed to get Metal queue")

        // Track if scene needs redraw (atomic for thread safety)
        val needsRedraw = AtomicBoolean(true)

        // Pending resize: holds resize event when resize is requested
        val pendingResize = AtomicReference<InputEvent?>(null)

        // Pending surface ID: set when we receive a new surface ID
        val pendingSurfaceID = AtomicReference<Int?>(null)

        // Event queue for input events
        val eventQueue = ConcurrentLinkedQueue<InputEvent>()

        // Start receiving events from host via socket
        // The initial surface ID will arrive as EVENT_TYPE_SURFACE_ID
        ipc.startReceiving(
            onInputEvent = { event ->
                if (event.type == InputType.RESIZE) {
                    // Resize events handled specially - store for main loop
                    pendingResize.set(event)
                } else {
                    eventQueue.offer(event)
                }
                needsRedraw.set(true)
            },
            onSurfaceID = { surfaceID ->
                pendingSurfaceID.set(surfaceID)
                needsRedraw.set(true)
            },
            onJuceEvent = onJuceEvent
        )

        // Wait for initial surface ID from host
        var initialSurfaceID: Int? = null
        while (initialSurfaceID == null && ipc.isRunning) {
            initialSurfaceID = pendingSurfaceID.getAndSet(null)
            if (initialSurfaceID == null) {
                Thread.sleep(10)
            }
        }

        if (initialSurfaceID == null) {
            error("Failed to receive initial surface ID")
        }

        // Initial render resources
        var resources = createRenderResources(metalContext, devicePtr, queuePtr, initialSurfaceID)

        // Track current scale factor (may change on resize if window moves to different display)
        var currentScale = scaleFactor

        // Create Compose scene - use scaleFactor for proper Retina/HiDPI rendering
        var scene = CanvasLayersComposeScene(
            density = Density(currentScale),
            size = IntSize(resources.width, resources.height),
            coroutineContext = Dispatchers.Unconfined,
            invalidate = { needsRedraw.set(true) }
        )
        scene.setContent(content)

        // Input dispatcher - needs scale factor to convert host points to Compose pixels
        var inputDispatcher = InputDispatcher(scene, currentScale)

        try {
            // Render loop - Compose draws directly to IOSurface!
            runBlocking {
                var frameCount = 0

                // Run until socket closes (host signals exit by closing socket)
                while (ipc.isRunning) {
                    try {
                        val frameStart = System.nanoTime()

                        // Check for pending resize + new surface ID
                        val resizeEvent = pendingResize.getAndSet(null)
                        val newSurfaceID = pendingSurfaceID.getAndSet(null)

                        if (resizeEvent != null && newSurfaceID != null) {
                            val newWidth = resizeEvent.width
                            val newHeight = resizeEvent.height
                            val newScale = resizeEvent.scaleFactor

                            // Close old resources (but keep the scene!)
                            resources.close()

                            // Create new resources for the new surface
                            resources = createRenderResources(metalContext, devicePtr, queuePtr, newSurfaceID)

                            // Update scene size - preserves all Compose state!
                            scene.size = IntSize(newWidth, newHeight)

                            // Update density if scale factor changed
                            if (newScale != currentScale) {
                                currentScale = newScale
                                scene.close()
                                scene = CanvasLayersComposeScene(
                                    density = Density(currentScale),
                                    size = IntSize(newWidth, newHeight),
                                    coroutineContext = Dispatchers.Unconfined,
                                    invalidate = { needsRedraw.set(true) }
                                )
                                scene.setContent(content)
                                inputDispatcher = InputDispatcher(scene, currentScale)
                            } else {
                                inputDispatcher.scaleFactor = currentScale
                            }

                            needsRedraw.set(true)
                        }

                        // Process pending input events
                        while (true) {
                            val event = eventQueue.poll() ?: break
                            inputDispatcher.dispatch(event)
                        }

                        // Render Compose content to IOSurface
                        val canvas = resources.skiaSurface.canvas
                        scene.render(canvas.asComposeCanvas(), frameStart)

                        // Flush and SYNC - this waits for GPU to finish
                        resources.skiaSurface.flushAndSubmit(syncCpu = true)

                        // Invoke frame callback if provided
                        onFrameRendered?.invoke(frameCount.toLong(), resources.skiaSurface)

                        if (frameCount == 0) {
                            ipc.sendFirstFrameRendered()
                        }
                        frameCount++
                        needsRedraw.set(false)
                    } catch (e: CancellationException) {
                        throw e
                    } catch (e: Exception) {
                        delay(100)
                    }
                }
            }
        } finally {
            ipc.stopReceiving()
            scene.close()
            resources.close()
        }
    } finally {
        NativeLib.INSTANCE.destroyMetalContext(metalContext)
    }
}

File: juce_cmp_ui/lib/src/main/cpp/iosurface_renderer.m
// SPDX-FileCopyrightText: 2026 Luciano Iam <oss@lucianoiam.com>
// SPDX-License-Identifier: MIT

#import <stdio.h>
#import <stdlib.h>
#import <unistd.h>
#import <IOSurface/IOSurface.h>
#import <Metal/Metal.h>

/**
 * Zero-copy Metal renderer for Compose IOSurface integration.
 *
 * This library provides the Metal device, command queue, and IOSurface-backed
 * texture that Skia can render to directly via DirectContext.makeMetal() and
 * BackendRenderTarget.makeMetal().
 *
 * Architecture:
 * - Kotlin creates a Metal context via createMetalContext()
 * - Kotlin creates an IOSurface-backed texture via createIOSurfaceTexture()
 * - Skia's DirectContext and BackendRenderTarget use these Metal resources
 * - Compose renders directly to the IOSurface - zero CPU pixel copies!
 */

// Metal context holding device and queue
typedef struct {
    id<MTLDevice> device;
    id<MTLCommandQueue> commandQueue;
} MetalContext;

// Create Metal context for GPU operations
void* createMetalContext(void) {
    @autoreleasepool {
        id<MTLDevice> device = MTLCreateSystemDefaultDevice();
        if (device == nil) {
            return NULL;
        }

        id<MTLCommandQueue> commandQueue = [device newCommandQueue];
        if (commandQueue == nil) {
            return NULL;
        }

        MetalContext* ctx = (MetalContext*)malloc(sizeof(MetalContext));
        ctx->device = device;
        ctx->commandQueue = commandQueue;

        // Prevent ARC from releasing these
        CFRetain((__bridge CFTypeRef)device);
        CFRetain((__bridge CFTypeRef)commandQueue);

        return ctx;
    }
}

// Destroy Metal context
void destroyMetalContext(void* context) {
    if (context == NULL) return;

    @autoreleasepool {
        MetalContext* ctx = (MetalContext*)context;

        CFRelease((__bridge CFTypeRef)ctx->commandQueue);
        CFRelease((__bridge CFTypeRef)ctx->device);

        free(ctx);
    }
}

// Get the MTLDevice pointer for Skia's DirectContext.makeMetal()
void* getMetalDevice(void* context) {
    if (context == NULL) return NULL;
    MetalContext* ctx = (MetalContext*)context;
    return (__bridge void*)ctx->device;
}

// Get the MTLCommandQueue pointer for Skia's DirectContext.makeMetal()
void* getMetalQueue(void* context) {
    if (context == NULL) return NULL;
    MetalContext* ctx = (MetalContext*)context;
    return (__bridge void*)ctx->commandQueue;
}

// Create an IOSurface-backed Metal texture for Skia's BackendRenderTarget.makeMetal()
// Returns the MTLTexture pointer that can be used with BackendRenderTarget.makeMetal()
void* createIOSurfaceTexture(void* context, int surfaceID, int* outWidth, int* outHeight) {
    if (context == NULL) return NULL;

    @autoreleasepool {
        MetalContext* ctx = (MetalContext*)context;

        // Lookup IOSurface by global ID
        IOSurfaceRef surface = IOSurfaceLookup((IOSurfaceID)surfaceID);
        if (surface == NULL) {
            return NULL;
        }

        size_t width = IOSurfaceGetWidth(surface);
        size_t height = IOSurfaceGetHeight(surface);

        if (outWidth) *outWidth = (int)width;
        if (outHeight) *outHeight = (int)height;

        // Create texture descriptor for IOSurface-backed texture
        MTLTextureDescriptor* textureDescriptor = [[MTLTextureDescriptor alloc] init];
        textureDescriptor.width = width;
        textureDescriptor.height = height;
        textureDescriptor.pixelFormat = MTLPixelFormatBGRA8Unorm;
        textureDescriptor.usage = MTLTextureUsageRenderTarget | MTLTextureUsageShaderRead;
        textureDescriptor.storageMode = MTLStorageModeShared;

        // Create texture backed by the IOSurface - this is the zero-copy magic!
        id<MTLTexture> texture = [ctx->device newTextureWithDescriptor:textureDescriptor
                                                             iosurface:surface
                                                                 plane:0];
        CFRelease(surface);

        if (texture == nil) {
            return NULL;
        }

        // Retain the texture so it survives autorelease
        CFRetain((__bridge CFTypeRef)texture);

        return (__bridge void*)texture;
    }
}

// Release an IOSurface-backed texture
void releaseIOSurfaceTexture(void* texturePtr) {
    if (texturePtr == NULL) return;

    @autoreleasepool {
        id<MTLTexture> texture = (__bridge id<MTLTexture>)texturePtr;
        CFRelease((__bridge CFTypeRef)texture);
    }
}

// Flush pending GPU work (call after Skia renders)
void flushAndSync(void* context) {
    if (context == NULL) return;

    @autoreleasepool {
        MetalContext* ctx = (MetalContext*)context;

        // Create a command buffer just to synchronize
        id<MTLCommandBuffer> commandBuffer = [ctx->commandQueue commandBuffer];
        [commandBuffer commit];
        [commandBuffer waitUntilCompleted];
    }
}

// Read data from a socket file descriptor
// Returns number of bytes read, or -1 on error
ssize_t socketRead(int socketFD, void* buffer, size_t length) {
    return read(socketFD, buffer, length);
}

// Write data to a socket file descriptor
// Returns number of bytes written, or -1 on error
ssize_t socketWrite(int socketFD, const void* buffer, size_t length) {
    return write(socketFD, buffer, length);
}
